src/v/kafka/client/direct_consumer/direct_consumer.cc

```diff
@@ -0,0 +1,186 @@
+    } catch (ss::condition_variable_timed_out&) {
+        co_return chunked_vector<fetched_topic_data>{};
+    }
+}
+
+void direct_consumer::update_configuration(configuration cfg) { _config = cfg; }
+
+ss::future<> direct_consumer::update_fetchers() {
+    // Stop fetch loops by acquiring a write lock
+    auto u = co_await _fetchers_lock.hold_write_lock();
+    for (auto& [topic, partitions] : _removals) {
+        for (auto& p : partitions) {
+            for (auto& [id, fetcher] : _broker_fetchers) {
+                // Unassign partitions from all fetchers
+                co_await fetcher->unassign_partition(
+                  model::topic_partition_view(topic, p));
+            }
+        }
+    }
+    for (auto& [topic, partitions] : _subscriptions) {
+        for (auto& [p_id, sub] : partitions) {
+            auto leader_id = _cluster->get_topics().leader(
+              model::topic_partition_view(topic, p_id));
+
+            if (!leader_id) {
+                ssx::spawn_with_gate(_gate, [this] {
+                    return _cluster->request_metadata_update();
+                });
+                if (sub.current_fetcher) {
+                    // If there is a current fetcher, unassign the partition
+                    auto& current = get_fetcher(*sub.current_fetcher);
+                    co_await current.unassign_partition(
+                      model::topic_partition_view(topic, p_id));
+                }
+                continue;
+            }
+
+            if (sub.current_fetcher != leader_id) {
+                // If the fetcher is not the same as the current one, we need to
+                // assign it
+                if (sub.current_fetcher) {
+                    auto& current = get_fetcher(*sub.current_fetcher);
+                    // If there was a previous fetcher, unassign it
+                    auto offset = co_await current.unassign_partition(
+                      model::topic_partition_view(topic, p_id));
+                    sub.fetch_offset = offset;
+                }
+                sub.current_fetcher = *leader_id;
+                auto& new_fetcher = get_fetcher(*leader_id);
+                co_await new_fetcher.assign_partition(
+                  model::topic_partition_view(topic, p_id), sub.fetch_offset);
+
+            } else if (sub.fetch_offset) {
+                // If the fetch offset is set, update it
+                auto& current = get_fetcher(*sub.current_fetcher);
+                co_await current.assign_partition(
+                  model::topic_partition_view(topic, p_id), sub.fetch_offset);
+            }
+            sub.fetch_offset.reset();
+        }
+    }
+    for (auto it = _broker_fetchers.begin(); it != _broker_fetchers.end();) {
+        auto& [id, fetcher] = *it;
+        if (fetcher->is_idle()) {
+            co_await fetcher->stop();
+            it = _broker_fetchers.erase(it);
+        } else {
+            ++it;
+        }
+    }
```

Comment:
was wondering if this code needs to reconcile brokers? (eg: metadata response no longer has a broker etc).